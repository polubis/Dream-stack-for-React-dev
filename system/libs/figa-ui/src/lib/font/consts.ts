const FONT_VARIANTS = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'b1',
  'b2',
  'b3',
] as const;
const FONT_MOTIVES = ['default', 'primary'] as const;

export { FONT_VARIANTS, FONT_MOTIVES };
