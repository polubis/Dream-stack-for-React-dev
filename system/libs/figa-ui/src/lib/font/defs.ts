import type { DetailedHTMLProps, HTMLAttributes, ReactNode } from 'react';
import type { FONT_MOTIVES, FONT_VARIANTS } from './consts';

type FontVariant = (typeof FONT_VARIANTS)[number];
type FontMotive = (typeof FONT_MOTIVES)[number];

type SupportedFontElement =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'span'
  | 'b'
  | 'strong'
  | 'i'
  | 'p'
  | 'em';

type VariantElementMap = Record<FontVariant, SupportedFontElement>;

type FontHTMLElementProps = Omit<
  DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>,
  'children'
>;

interface FontProps extends FontHTMLElementProps {
  element?: SupportedFontElement;
  variant: FontVariant;
  children: ReactNode;
  italic?: boolean;
  trim?: boolean;
  motive?: FontMotive;
  bold?: boolean;
}

export type {
  FontProps,
  FontVariant,
  SupportedFontElement,
  VariantElementMap,
  FontMotive,
};
