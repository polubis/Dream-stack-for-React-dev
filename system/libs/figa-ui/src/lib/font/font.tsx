import type { FontProps, VariantElementMap } from './defs';

import { createElement } from 'react';
import c from 'classnames';

const FONT_VARIANT_ELEMENT_MAP: VariantElementMap = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  b1: 'p',
  b2: 'p',
  b3: 'p',
};

const Font = ({
  className,
  variant,
  element,
  italic,
  motive = 'default',
  trim,
  bold,
  ...props
}: FontProps) => {
  return createElement(element ?? FONT_VARIANT_ELEMENT_MAP[variant], {
    className: c(
      'font',
      variant,
      motive,
      className,
      { italic },
      { bold },
      { trim }
    ),
    ...props,
  });
};

export { Font };
