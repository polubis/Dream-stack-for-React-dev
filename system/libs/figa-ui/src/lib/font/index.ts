export * from './font';
export * from './defs';
