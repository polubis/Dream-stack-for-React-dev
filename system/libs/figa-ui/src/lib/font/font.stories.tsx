import type { Story, Meta } from '@storybook/react';

import { Font } from './font';
import { FONT_MOTIVES, FONT_VARIANTS } from './consts';
import { Box } from '../box';

export default {
  component: Font,
  title: 'Font',
} as Meta;

const Template: Story = () => {
  return (
    <Box padding={[300, 300, 300, 300]} spacing={[200, 200, 200, 200]}>
      <Box>
        {FONT_VARIANTS.map((variant) => (
          <div key={variant}>
            <Font variant={variant} children={variant} />
          </div>
        ))}
      </Box>

      <Box>
        {FONT_VARIANTS.map((variant) => (
          <div key={'italic' + variant}>
            <Font italic variant={variant} children={variant} />
          </div>
        ))}
      </Box>

      <Box>
        {FONT_VARIANTS.map((variant) => (
          <div key={'bold' + variant}>
            <Font variant={variant} children="Some text" bold />
          </div>
        ))}
      </Box>

      <Box>
        {FONT_MOTIVES.map((motive) => (
          <div key={'motive' + motive}>
            <Font variant="b1" motive={motive} children={motive} />
          </div>
        ))}
      </Box>

      <Box maxWidth="80px">
        {FONT_MOTIVES.map((motive) => (
          <div key={'motive' + motive}>
            <Font variant="h3" trim motive={motive} children={motive} />
          </div>
        ))}
      </Box>
    </Box>
  );
};

export const Default = Template.bind({});
Default.args = {};
