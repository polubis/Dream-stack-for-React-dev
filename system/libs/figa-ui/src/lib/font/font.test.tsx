import { fireEvent, render, screen } from '@testing-library/react';
import { Font } from './font';

describe('Font can be used when: ', () => {
  it('[FRAGILE] assigns classes', () => {
    const { asFragment } = render(
      <Font className="my-class" variant="h5" bold italic trim>
        My content inside
      </Font>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('[FRAGILE] allows to use other than default element', () => {
    const { container } = render(
      <Font variant="h5" element="span">
        My content inside
      </Font>
    );

    const font = container.querySelector('.font');

    expect(font?.tagName).toBe('SPAN');
  });

  it('allows to pass any other properties', () => {
    const spy = jest.fn();

    render(
      <Font variant="h5" onClick={spy}>
        My content inside
      </Font>
    );

    fireEvent.click(screen.getByText(/My content inside/));

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
