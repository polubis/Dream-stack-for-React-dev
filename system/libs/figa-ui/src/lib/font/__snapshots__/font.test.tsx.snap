// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Font can be used when:  [FRAGILE] assigns classes 1`] = `
<DocumentFragment>
  <h5
    class="font h5 default my-class italic bold trim"
  >
    My content inside
  </h5>
</DocumentFragment>
`;
